lib/cookies.js

```diff
@@ -0,0 +1,97 @@
+const { CEILING_PRIORITY } = require('./trackerAllowlist')
+const { generateTrackerDomainAliasMapping, getTrackerEntryDomain, storeInLookup } = require('./utils')
+
+const COOKIE_PRIORITY = CEILING_PRIORITY + 100
+
+/**
+ * Build HTTP cookie blocking rules from the blocklist and cookie configuration
+ * @param {import('./utils').TDS} tds The tracker blocklist
+ * @param {string[]} excludedCookieDomains 3p domains for which we shouldn't block cookies
+ * @param {string[]} siteAllowlist Sites on which we shouldn't block 3p cookies
+ * @returns {import('./utils').DNRRule[]} Cookie blocking rules
+ */
+function generateCookieBlockingRuleset (tds, excludedCookieDomains, siteAllowlist, startingRuleId = 1) {
+    /**
+     * @type {import('./utils').DNRRule[]}
+     */
+    const rules = []
+    /**
+     * @type {Map<string, { domains: Set<string>, trackerDomains: Set<string> }>}
+     */
+    const entityDomainMapping = new Map()
+    /**
+     * @type {Map<string, string[]>}
+     */
+    const trackerDomainExclusions = new Map()
+    const singleDomainEntityDomains = []
+    // collect CNAMEs for each tracker
+    const requestDomainsByTrackerDomain = generateTrackerDomainAliasMapping(tds)
+
+    // process exclusions: find their tracker domain so we know which rule to put it in
+    excludedCookieDomains.forEach(d => {
+        const trackerEntryDomain = getTrackerEntryDomain(tds.trackers, d)
+        storeInLookup(trackerDomainExclusions, trackerEntryDomain, [d])
+    })
+
+    // Gather trackers by owner and build the set of owned and owned tracker domains for each
+    for (const [trackerDomain, trackerEntry] of Object.entries(tds.trackers)) {
```

Comment:
How about just building the `trackerDomainsByEntity` lookup here? That would make the logic a bit simpler (no need for nested lookups). I figure we can just refer to  `tds.entities[trackerEntry.owner.name].domains` directly when we need it, or am I missing something?

Reply:
Unfortunately not. While `tds.trackers` entries are _generally_ eTLD+1, there are some exceptions (e.g. `bad.third-party.site`). Likewise, `tds.entities[trackerEntry.owner.name].domains` are _generally_ eTLD+1, some are no (e.g. `cloudfront.net`). This means that we cannot just gather the trackers stated in the latter because we'll miss some.

Reply:
I'm too tired today (or too stupid generally - you pick) to grok this, but I'll take your word for it. You've clearly thought it through :+1: .